common/lib/client/realtimechannel.ts

```diff
@@ -1,27 +1,84 @@
 import ProtocolMessage from '../types/protocolmessage';
 import EventEmitter from '../util/eventemitter';
-import Utils from '../util/utils';
+import * as Utils from '../util/utils';
 import Channel from './channel';
 import Logger from '../util/logger';
 import RealtimePresence from './realtimepresence';
-import Message from '../types/message';
+import Message, { CipherOptions } from '../types/message';
 import ChannelStateChange from './channelstatechange';
 import ErrorInfo from '../types/errorinfo';
 import PresenceMessage from '../types/presencemessage';
-import ConnectionError from '../transport/connectionerror';
-
-var RealtimeChannel = (function() {
-	var actions = ProtocolMessage.Action;
-	var noop = function() {};
-	var statechangeOp = 'statechange';
-	var syncOp = 'sync';
-
-	/* public constructor */
-	function RealtimeChannel(realtime, name, options) {
+import ConnectionErrors from '../transport/connectionerrors';
+import * as API from '../../../ably';
+import ConnectionManager from '../transport/connectionmanager';
+import ConnectionStateChange from './connectionstatechange';
+import { ErrCallback, PaginatedResultCallback } from '../../types/utils';
+
+// TODO: Replace this with the real type when Realtime is in TypeScript
+type Realtime = any;
+
+export type RealtimeChannelState = 'initialized' | 'attaching' | 'attached' | 'suspended' | 'detaching' | 'detached' | 'failed';
```

Comment:
There seems to now be duplication in a lot of places now between types defined in ably.d.ts and in the code like this one (which looks like it should be equivalent to `API.Types.ChannelState`). This makes it quite fragile against future changes - would be better defined in one place and referenced in the other. (Doesn't strictly matter which, but might be nicer for future maintainability to have it be defined in the code itself and just imported by ably.d.ts?)

Reply:
We started using API.Types while this change was still in review so it slipped through, I've fixed it now though